flutter-studio/src/io/flutter/FlutterStudioStartupActivity.java

```diff
@@ -109,4 +132,62 @@ public void sourceGenerationFinished(@NotNull Project project) {
       }
     };
   }
+
+  @NotNull
+  private static DebuggerManagerListener makeAddToAppAttachListener(@NotNull Project project) {
+
+    return new DebuggerManagerListener() {
+
+      DebugProcessListener dpl = new DebugProcessListener() {
+        @Override
+        public void processDetached(@NotNull DebugProcess process, boolean closedByUser) {
+          ThreeState state = project.getUserData(ATTACH_IS_ACTIVE);
+          if (state != null) {
+            project.putUserData(ATTACH_IS_ACTIVE, null);
+          }
+        }
+
+        @Override
+        public void processAttached(@NotNull DebugProcess process) {
+          if (project.getUserData(ATTACH_IS_ACTIVE) != null) {
+            return;
+          }
+          // Launch flutter attach if a run config can be found.
+          if (findRunConfig(project) == null) {
+            // Either there is no Flutter run config or there are more than one.
+            if (RunManagerEx.getInstanceEx(project).getSelectedConfiguration() instanceof SdkRunConfig) {
+              // The selected run config at this point is not Flutter, so we can't start the process automatically.
+              return;
+            }
+          }
+          FlutterSdk sdk = FlutterSdk.getFlutterSdk(project);
+          if (sdk == null) {
+            return;
+          }
+          // If needed, DataContext could be saved by FlutterReloadManager.beforeActionPerformed() in project user data.
+          DataContext context = DataContext.EMPTY_CONTEXT;
+          List<Module> modules = FlutterModuleUtils.findModulesWithFlutterContents(project);
+          assert modules.size() == 1; // TODO(messick) Need to change this if multiple :flutter sub-projects supported.
+          PubRoot pubRoot = PubRoot.forDirectory(Objects.requireNonNull(modules.get(0).getModuleFile()).getParent());
+          Application app = ApplicationManager.getApplication();
+          project.putUserData(ATTACH_IS_ACTIVE, ThreeState.fromBoolean(true));
+          // Note: Using block comments to preserve formatting.
+          app.invokeLater( /* After the Android launch completes, */
+            () -> app.executeOnPooledThread( /* but not on the EDT, */
+              () -> app.runReadAction( /* with read access, */
+                () -> new AttachDebuggerAction().startCommand(project, sdk, pubRoot, context)))); /* attach. */
+        }
+      };
+
+      @Override
+      public void sessionCreated(DebuggerSession session) {
+        session.getProcess().addDebugProcessListener(dpl);
+      }
+
+      @Override
+      public void sessionRemoved(DebuggerSession session) {
+        session.getProcess().removeDebugProcessListener(dpl);
```

Comment:
Does this kill the attach process? Could it ever leak now that attach stays alive after the observatory is closed?

Reply:
@blasten This runs when the Android app exits. At that point, my experiments all showed the attach process had already terminated. I don't think the observatory is involved here, but I honestly don't know for sure.